File: 238-product-of-array-except-self/product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> leftMul(n);
        vector<int> rightMul(n);

        leftMul[0] = nums[0];
        for(int i = 1; i < n; ++i) {
            leftMul[i] = nums[i] * leftMul[i - 1];
        }

        rightMul[n - 1] = nums[n - 1];
        for(int i = n - 2; i >= 0; i--) {
            rightMul[i] = nums[i] * rightMul[i + 1];
        }

        vector<int> res(n);
        for(int i = 0; i < n; i++) {
            if (i == 0) {
                res[i] = rightMul[1];
            } else if (i == n - 1) {
                res[i] = leftMul[n - 2];
            } else {
                res[i] = rightMul[i + 1] * leftMul[i - 1];
            }
        }

        return res;
    }
};

File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        left = [1] * n
        right = [1] * n

        for i in range(n):
            if i == 0:
                left[i] = nums[i]
            else:
                left[i] = nums[i] * left[i-1]
        
        for i in range(n - 1, -1, -1):
            if i == n - 1:
                right[i] = nums[i]
            else:
                right[i] = nums[i] * right[i+1]
        
        res = [1] * n
        for i in range(n):
            if i == 0:
                res[i] = right[i+1]
            elif i == n-1:
                res[i] = left[i-1]
            else:
                res[i] = left[i-1] * right[i+1]
                
        return res
